reactos/drivers/usb/usbport/device.c

```diff
@@ -0,0 +1,1873 @@
+    {
+        DeviceHandle->DeviceSpeed = UsbLowSpeed;
+    }
+    else if ( PortStatus & USB_PORT_STATUS_HIGH_SPEED )
+    {
+        DeviceHandle->DeviceSpeed = UsbHighSpeed;
+    }
+    else
+    {
+        DeviceHandle->DeviceSpeed = UsbFullSpeed;
+    }
+
+    KeWaitForSingleObject(&FdoExtension->DeviceSemaphore,
+                          Executive,
+                          KernelMode,
+                          FALSE,
+                          NULL);
+
+    PipeHandle = &DeviceHandle->PipeHandle;
+
+    PipeHandle->Flags = PIPE_HANDLE_FLAG_CLOSED;
+
+    PipeHandle->EndpointDescriptor.bLength = sizeof(PipeHandle->EndpointDescriptor);
+    PipeHandle->EndpointDescriptor.bDescriptorType = USB_ENDPOINT_DESCRIPTOR_TYPE;
+
+    if (DeviceHandle->DeviceSpeed == UsbLowSpeed)
+    {
+        PipeHandle->EndpointDescriptor.wMaxPacketSize = 8;
+    }
+    else
+    {
+        PipeHandle->EndpointDescriptor.wMaxPacketSize = USB_DEFAULT_MAX_PACKET;
+    }
+
+    InitializeListHead(&DeviceHandle->PipeHandleList);
+
+    Status = USBPORT_OpenPipe(FdoDevice,
+                              DeviceHandle,
+                              PipeHandle,
+                              NULL);
+
+    if (NT_SUCCESS(Status))
+    {
+        IsOpenedPipe = TRUE;
+    }
+
+    if (NT_ERROR(Status))
+    {
+        DPRINT1("USBPORT_CreateDevice: USBPORT_OpenPipe return - %p\n", Status);
+
+        KeReleaseSemaphore(&FdoExtension->DeviceSemaphore,
+                           LOW_REALTIME_PRIORITY,
+                           1,
+                           FALSE);
+
+        ExFreePool(DeviceHandle);
+
+        return Status;
+    }
+
+    DeviceDescriptor = ExAllocatePoolWithTag(NonPagedPool,
+                                             USB_DEFAULT_MAX_PACKET,
+                                             USB_PORT_TAG);
+
+    if (!DeviceDescriptor)
+    {
+        DPRINT1("USBPORT_CreateDevice: Not allocated DeviceDescriptor\n");
+        goto ErrorExit;
+    }
+
+    RtlZeroMemory(DeviceDescriptor, USB_DEFAULT_MAX_PACKET);
+    RtlZeroMemory(&SetupPacket, sizeof(USB_DEFAULT_PIPE_SETUP_PACKET));
+
+    SetupPacket.bmRequestType._BM.Dir = BMREQUEST_DEVICE_TO_HOST;
+    SetupPacket.bRequest = USB_REQUEST_GET_DESCRIPTOR;
+    SetupPacket.wValue.HiByte = USB_DEVICE_DESCRIPTOR_TYPE;
+    SetupPacket.wLength = USB_DEFAULT_MAX_PACKET;
+
+    TransferedLen = 0;
+
+    Status = USBPORT_SendSetupPacket(DeviceHandle,
+                                     FdoDevice,
+                                     &SetupPacket,
+                                     DeviceDescriptor,
+                                     USB_DEFAULT_MAX_PACKET,
+                                     &TransferedLen,
+                                     NULL);
+
+    RtlCopyMemory(&DeviceHandle->DeviceDescriptor,
+                  DeviceDescriptor,
+                  sizeof(USB_DEVICE_DESCRIPTOR));
+
+    ExFreePool(DeviceDescriptor);
+
+    if ((TransferedLen == 8) && !NT_SUCCESS(Status))
```

Comment:
That makes sense. But the code should express the logic behind this -- RTL_SIZEOF_THROUGH_FIELD(USB_DEVICE_DESCRIPTOR, bMaxPacketSize0)